protonvpn_linux_gui/presenters/dashboard_presenter.py

```diff
@@ -0,0 +1,565 @@
+
+        gui_logger.debug(">>> Result: \"{0}\"".format(result))
+        
+        gobject.idle_add(self.update_labels_status, update_labels_dict)
+
+        gui_logger.debug(">>> Ended tasks in \"fastest\" thread.")
+
+    def on_last_connect(self, **kwargs):
+        """Function that connects to the last connected server.
+        """        
+        gui_logger.debug(">>> Running \"reconnect\".")
+
+        update_labels_dict = {
+            "connection_labels": kwargs.get("connection_labels"),
+            "servers": False,
+            "disconnecting": False,
+            "conn_info": False
+        }
+
+        result = self.dashboard_service.last_connect()
+
+        server_protocol = get_server_protocol_from_cli(result, return_protocol=True)
+
+        display_message = result.stdout.decode()
+
+        if server_protocol:
+            display_message = "You are connected to <b>{}</b> via <b>{}</b>!".format(server_protocol[0], server_protocol[1].upper())
+
+        self.queue.put(dict(action="update_dialog", label=display_message))
+
+        gui_logger.debug(">>> Result: \"{0}\"".format(result))
+
+        gobject.idle_add(self.update_labels_status, update_labels_dict)
+
+        gui_logger.debug(">>> Ended tasks in \"reconnect\" thread.")
+
+    def random_connect(self, **kwargs):
+        """Function that connects to a random server.
+        """
+        gui_logger.debug(">>> Running \"reconnect\"")
+
+        update_labels_dict = {
+            "connection_labels": kwargs.get("connection_labels"),
+            "servers": False,
+            "disconnecting": False,
+            "conn_info": False
+        }
+
+        result = self.dashboard_service.random_connect()
+
+        display_message = result.stdout.decode()
+        server_protocol = get_server_protocol_from_cli(result, return_protocol=True)
+
+        if server_protocol:
+            display_message = "You are connected to <b>{}</b> via <b>{}</b>!".format(server_protocol[0], server_protocol[1].upper())
+
+        self.queue.put(dict(action="update_dialog", label=display_message))
+
+        gui_logger.debug(">>> Result: \"{0}\"".format(result))
+
+        gobject.idle_add(self.update_labels_status, update_labels_dict)
+
+        gui_logger.debug(">>> Ended tasks in \"random_c\" thread.")
+
+    def on_refresh_servers(self, **kwargs):
+        """Function that reloads server list to either secure-core or non-secure-core.
+        """  
+        # Sleep is needed because it takes a second to update the information,
+        # which makes the button "lag". Temporary solution.
+        time.sleep(1)
+        gui_logger.debug(">>> Running \"update_reload_secure_core_serverslabels_server_list\".")
+
+        return_val = False
+        populate_servers_dict = {
+            "tree_object": kwargs.get("tree_object"),
+            "servers": False
+        }
+
+        self.queue.put(dict(action="hide_spinner"))
+
+        conn = custom_get_ip_info()
```

Comment:
Sounds like this should a general property that is always kept updated?

Reply:
The queue actually calls a function in a thread-safe manner. This is made this way for abstraction so that an instance of message_dialog is not passed, while making possible inter-thread message exchange possible.